alerts/resource_alerts.libsonnet

```diff
@@ -34,18 +34,34 @@
           } +
           if $._config.showMultiCluster then {
             expr: |||
-              sum(namespace_cpu:kube_pod_container_resource_requests:sum{%(ignoringOverprovisionedWorkloadSelector)s}) by (%(clusterLabel)s) - (sum(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s) - max(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s)) > 0
+              (count(kube_node_info) == 1
               and
-              (sum(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s) - max(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s)) > 0
+              sum(namespace_cpu:kube_pod_container_resource_requests:sum{%(ignoringOverprovisionedWorkloadSelector)s}) by (%(clusterLabel)s) -
+              sum(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s) > 0)
+              or
+              (sum(namespace_cpu:kube_pod_container_resource_requests:sum{%(ignoringOverprovisionedWorkloadSelector)s}) by (%(clusterLabel)s) -
+              (sum(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s) -
+              max(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s)) > 0
+              and
+              (sum(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s) -
+              max(kube_node_status_allocatable{%(kubeStateMetricsSelector)s,resource="cpu"}) by (%(clusterLabel)s)) > 0)
```

Comment:
From what I gather here, this pre-existing latter part of the `and` effectively removes single-node clusters (because "sum(cpu) - max(cpu)" can never be greater than zero on a single-node cluster).

This seems to conflict with the PR description:

> the alert will fire if any request limits exist

Can you be more specific or provide a test case where the existing rule is not behaving as expected? Maybe this particular alert isn't affected?

Reply:
👋🏼 So, the following hypothesis:

> the alert will fire if any request limits exist

is true when we reduce the expression by equating `sum(Q) by (cluster) = max(Q) by (cluster)`.

In the original expression,

```
sum(QQ) by (cluster) - (sum(Q) by (cluster) - max(Q) by (cluster)) > 0 and (sum(Q) by (cluster) - max(Q) by (cluster)) > 0
```

the latter part after the `and` is `false` for SNO, so the alert never fires, but please LMK if I'm missing something.

Reply:
Ok, so the problem is not that:

> the alert will fire if any request limits exist

The problem is:

> the latter part after the and is false for SNO, so the alert never fires

So TL;DR: the current alert only works for clusters with more than 1 node.

And reading back your desired solution: you want to have an alert which works for both single-node clusters and multi-node clusters?

Reply:
Exactly, that would be the desired outcome that I'm going for.

Reply:
ok great thanks - I'll take another look tomorrow